### PR TITLE
Rewrite Migrations

### PR DESCRIPTION
**Pull request by ghostbear**

### CODE REVIEW

File: app/src/main/java/mihon/core/migration/MigrationContext.kt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package mihon.core.migration
+
+import uy.kohesive.injekt.Injekt
+
+class MigrationContext {
+
+    inline fun <reified T> get(): T? {
+        return Injekt.getInstanceOrNull(T::class.java)
+    }
+
```
Comment (AntsyLich) [inline app/src/main/java/mihon/core/migration/MigrationContext.kt:10]:
```suggestion
```

File: app/src/test/java/mihon/core/migration/MigratorTest.kt [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            migrations = listOf(
+                Migration.of(Migration.ALWAYS) { true },
+                Migration.of(2f) { true },
+                Migration.of(3f) { true }),
+            onMigrationComplete = spyk
+        )
+        verify { spyk() }
+        Assertions.assertEquals(2, migrations.size)
+    }
+
```
Comment (AntsyLich) [inline app/src/test/java/mihon/core/migration/MigratorTest.kt:95]:
```suggestion
```

File: app/src/main/java/mihon/migration/TrustExtensionRepositoryMigration.kt [hunk 3]
```diff
@@ -0,0 +1,34 @@
+package mihon.migration
```
Comment (AntsyLich) [inline app/src/main/java/mihon/migration/TrustExtensionRepositoryMigration.kt:1]:
```suggestion
package mihon.core.migration
```

File: app/src/main/java/mihon/migration/SetupLibraryUpdateMigration.kt [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package mihon.migration
```
Comment (AntsyLich) [inline app/src/main/java/mihon/migration/SetupLibraryUpdateMigration.kt:1]:
```suggestion
package mihon.core.migration
```

File: app/src/main/java/mihon/migration/SetupBackupCreateMigration.kt [hunk 5]
```diff
@@ -0,0 +1,16 @@
+package mihon.migration
```
Comment (AntsyLich) [inline app/src/main/java/mihon/migration/SetupBackupCreateMigration.kt:1]:
```suggestion
package mihon.core.migration
```

File: app/src/main/java/mihon/migration/Migrations.kt [hunk 6]
```diff
@@ -0,0 +1,10 @@
+package mihon.migration
```
Comment (AntsyLich) [inline app/src/main/java/mihon/migration/Migrations.kt:1]:
```suggestion
package mihon.core.migration
```

File: app/src/main/java/mihon/core/migration/Migrator.kt [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    context (CoroutineScope)
+    private fun MigrationContext.migrate(versions: List<Int>, migrationsByVersion: Map<Int, List<Migration>>, dryrun: Boolean): Array<Deferred<Boolean>> {
+        return versions.sorted()
+            .flatMap { version ->
+                (migrationsByVersion[version] ?: emptyList())
+                    .sortedBy(Migration::version)
+                    .map { migration ->
+                        if (dryrun.not()) {
```
Comment (AntsyLich) [inline app/src/main/java/mihon/core/migration/Migrator.kt:43]:
```suggestion
                        if (!dryrun) {
```

File: app/src/main/java/mihon/core/migration/Migrator.kt [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+                            logcat { "Running migration: { name = ${migration::class.simpleName}, version = ${migration.version} }" }
+                            return@map async { migration.action(this@migrate) }
+                        }
+                        logcat { "(Dry-run) Running migration: { name = ${migration::class.simpleName}, version = ${migration.version} }" }
+                        async { true }
+                    }
+            }.toTypedArray()
```
Comment (AntsyLich) [inline app/src/main/java/mihon/core/migration/Migrator.kt:50]:
```suggestion
            }
                .toTypedArray()
```

File: app/src/main/java/eu/kanade/tachiyomi/ui/main/MainActivity.kt [hunk 9]
```diff
@@ -1,5 +1,6 @@
 package eu.kanade.tachiyomi.ui.main
 
+import androidx.compose.ui.graphics.Color.Companion as ComposeColor
```
Comment (AntsyLich) [inline app/src/main/java/eu/kanade/tachiyomi/ui/main/MainActivity.kt:3]: Why is this at top :kannawhat:

**Review by AntsyLich — CHANGES_REQUESTED** [review]

File: app/src/main/java/eu/kanade/tachiyomi/ui/main/MainActivity.kt (hunk 9 above)
Comment (ghostbear) [reply app/src/main/java/eu/kanade/tachiyomi/ui/main/MainActivity.kt:3]: AS deciding to move it

File: app/src/main/java/mihon/migration/SetupLibraryUpdateMigration.kt (hunk 4 above)
Comment (ghostbear) [reply app/src/main/java/mihon/migration/SetupLibraryUpdateMigration.kt:1]: Do we want the actual migrations to be in the same package as the Migrator logic?

Comment (AntsyLich) [reply app/src/main/java/mihon/migration/SetupLibraryUpdateMigration.kt:1]: can put them in a subpackage 🤔

File: app/src/main/java/mihon/core/migration/Migration.kt [hunk 10]
```diff
@@ -0,0 +1,19 @@
+package mihon.core.migration
+
+interface Migration {
+    val version: Float
```
Comment (AntsyLich) [inline app/src/main/java/mihon/core/migration/Migration.kt:4]: why float?

File: app/src/main/java/mihon/core/migration/Migrator.kt [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+object Migrator {
+
+    @SuppressWarnings("ReturnCount")
+    fun migrate(
+        old: Int,
+        new: Int,
+        migrations: List<Migration>,
+        dryrun: Boolean = false,
+        onMigrationComplete: () -> Unit
```
Comment (AntsyLich) [inline app/src/main/java/mihon/core/migration/Migrator.kt:14]: shouldn't this run after migration completes?

File: app/src/main/java/mihon/core/migration/Migrator.kt [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    @SuppressWarnings("MaxLineLength")
+    private fun MigrationContext.migrate(migrations: List<Migration>, dryrun: Boolean): Boolean {
+        return migrations.sortedBy { it.version }
+            .map { migration ->
+                if (!dryrun) {
+                    logcat { "Running migration: { name = ${migration::class.simpleName}, version = ${migration.version} }" }
+                } else {
+                    logcat { "(Dry-run) Running migration: { name = ${migration::class.simpleName}, version = ${migration.version} }" }
+                }
+                runBlocking { migration.action(this@migrate) }
```
Comment (AntsyLich) [inline app/src/main/java/mihon/core/migration/Migrator.kt:48]: shouldn't this be inside the if block?

File: app/src/main/java/mihon/core/migration/Migration.kt (hunk 10 above)
Comment (ghostbear) [reply app/src/main/java/mihon/core/migration/Migration.kt:4]: To allow multiple migrations for the same version and allow to choose if it should run after another

Comment (AntsyLich) [reply app/src/main/java/mihon/core/migration/Migration.kt:4]: Gigabrain

File: app/src/main/java/mihon/core/migration/Migrator.kt [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+        dryrun: Boolean = false,
+        onMigrationComplete: () -> Unit
+    ): Boolean {
+        val migrationContext = MigrationContext()
+
+        if (old == 0) {
+            return migrationContext.migrate(
+                migrations = migrations.filter { it.isAlways() },
+                dryrun = dryrun
+            ).also { onMigrationComplete() }
```
Comment (AntsyLich) [inline app/src/main/java/mihon/core/migration/Migrator.kt:22]:
```suggestion
            )
                .also { onMigrationComplete() }
```

File: app/src/main/java/mihon/core/migration/Migrator.kt [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+        }
+
+        if (old >= new) {
+            return false
+        }
+
+        return migrationContext.migrate(
+            migrations = migrations.filter { it.isAlways() || it.version.toInt() in (old + 1).. new },
+            dryrun = dryrun
+        ).also { onMigrationComplete() }
```
Comment (AntsyLich) [inline app/src/main/java/mihon/core/migration/Migrator.kt:32]:
```suggestion
        )
            .also { onMigrationComplete() }
```

File: app/src/main/java/mihon/core/migration/Migration.kt [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+
+    suspend fun action(migrationContext: MigrationContext): Boolean
```
Comment (AntsyLich) [inline app/src/main/java/mihon/core/migration/Migration.kt:6]: Thoughts on making this invoke operator fucntion so we can do `migration(this@migrate)`?

**Pull request merged**